File: sdk/alcedocore-sdk-python/alcedo_sdk/resources/roles.py
from __future__ import annotations

from typing import Any, Optional

from .base import BaseResource


class RolesResource(BaseResource):
    async def list(self) -> dict:
        return await self._request("GET", "/api/roles")

    async def get(self, id: str) -> dict:
        return await self._request("GET", f"/api/roles/{id}")

    async def create(self, data: dict) -> dict:
        return await self._request("POST", "/api/roles", json=data)

    async def update(self, id: str, data: dict) -> dict:
        return await self._request("PUT", f"/api/roles/{id}", json=data)

    async def delete(self, id: str) -> dict:
        return await self._request("DELETE", f"/api/roles/{id}")

    async def list_scopes(self, id: str) -> dict:
        return await self._request("GET", f"/api/roles/{id}/scopes")

    async def update_scopes(self, id: str, scopes: list[str]) -> dict:
        return await self._request("PUT", f"/api/roles/{id}/scopes", json={"scopes": scopes})

    async def list_policies(self, id: str) -> dict:
        return await self._request("GET", f"/api/roles/{id}/policies")

    async def assign_policy(self, id: str, policy_id: str) -> dict:
        return await self._request("POST", f"/api/roles/{id}/policies", json={"policy_id": policy_id})

    async def remove_policy(self, id: str, policy_id: str) -> dict:
        return await self._request("DELETE", f"/api/roles/{id}/policies/{policy_id}")

File: sdk/alcedocore-sdk-node/src/roles.ts
export function createRolesResource(ky: any) {
  return {
    list: (options?: any) => ky.get("app/roles", options).json(),
    get: (id: string, options?: any) => ky.get(`app/roles/${encodeURIComponent(id)}`, options).json(),
    create: (data: any, options?: any) => ky.post("app/roles", { json: data, ...options }).json(),
    update: (id: string, data: any, options?: any) => ky.put(`app/roles/${encodeURIComponent(id)}`, { json: data, ...options }).json(),
    delete: (id: string, options?: any) => ky.delete(`app/roles/${encodeURIComponent(id)}`, options).json(),
    listPermissions: (id: string, options?: any) => ky.get(`app/roles/${encodeURIComponent(id)}/permissions`, options).json(),
    updatePermissions: (id: string, data: any, options?: any) => ky.post(`app/roles/${encodeURIComponent(id)}/permissions`, { json: data, ...options }).json(),
    deletePermission: (id: string, permissionId: string, options?: any) => ky.delete(`app/roles/${encodeURIComponent(id)}/permissions/${encodeURIComponent(permissionId)}`, options).json(),
    listPolicies: (id: string, options?: any) => ky.get(`app/roles/${encodeURIComponent(id)}/policies`, options).json(),
    assignPolicy: (id: string, policyId: string, options?: any) => ky.post(`app/roles/${encodeURIComponent(id)}/policies`, { json: { policy_id: policyId }, ...options }).json(),
    removePolicy: (id: string, policyId: string, options?: any) => ky.delete(`app/roles/${encodeURIComponent(id)}/policies/${encodeURIComponent(policyId)}`, options).json(),
  };
}
